### PR TITLE
export setting CCC UI coding

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

**Comment by DhaaraniCIT** [comment]
Export Setting CCC section
Screenshot 2023-10-20 at 3 47 51 PM

File: src/app/core/services/sage300/sage300-helper/sage300-helper.service.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+})
+export class Sage300HelperService {
+
+  constructor(
+    private workspaceService: WorkspaceService,
+    private mappingService: Sage300MappingService,
+    private toastService: IntegrationsToastService
+  ) { }
+
+  importAttributes(isRefresh: boolean) {
```
Comment (ashwin1111) [inline src/app/core/services/sage300/sage300-helper/sage300-helper.service.ts:18]: return type

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 2]
```diff
@@ -120,11 +121,12 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
-    private formBuilder: FormBuilder
+    private formBuilder: FormBuilder,
+    private helperService: Sage300HelperService
   ) { }
 
-  getEmployeeFieldMapping(exportType: ExpenseGroupedBy): string {
-    return exportType[0].toUpperCase() + exportType.substr(1);
+  getFieldMapping(exportType: ExpenseGroupedBy): string {
+    return exportType[0].toUpperCase() + exportType.substr(1).toLowerCase();
   }
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:130]: move to helper, use angular pipe

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 3]
```diff
@@ -209,20 +211,34 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+        "reimbursable_expense_grouped_by": "EXPENSE",
+        "credit_card_expense_export_type": "PURCHASE_INVOICE",
+        "credit_card_expense_state": "CLOSED",
+        "credit_card_expense_grouped_by": "EXPENSE",
+        "credit_card_expense_date": "APPROVED_AT",
+        "default_ccc_account_name": "credit card account",
+        "default_ccc_account_id": "12312",
+        "default_vendor_name": "Nilesh",
+        "default_vendor_id": "123"
+      };
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:229]: let's use mock apis from start please

File: src/app/shared/components/configuration/configuration-step-header/configuration-step-header.component.html [hunk 4]
```diff
@@ -17,7 +17,7 @@ <h5 class="tw-text-faded-text-color tw-text-14-px tw-pt-8-px !tw-font-400">
           </h5>
         </div>
 
-        <button *ngIf="showSyncButton" [pTooltip]="'The integration will import all the newly updated Sage Intacct dimensions and Fyle expenses in the configured state of export'"
+        <button *ngIf="showSyncButton && !appName" [pTooltip]="'The integration will import all the newly updated Sage Intacct dimensions and Fyle expenses in the configured state of export'"
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-step-header/configuration-step-header.component.html:20]: why !appName when we have showSyncButton already?

File: src/app/shared/components/configuration/configuration-step-header/configuration-step-header.component.html [hunk 5]
```diff
@@ -26,5 +26,14 @@ <h5 class="tw-text-faded-text-color tw-text-14-px tw-pt-8-px !tw-font-400">
...
         </button>
 
+        <button *ngIf="showSyncButton && appName" [pTooltip]="'The integration will import all the newly updated Sage Intacct dimensions and Fyle expenses in the configured state of export'"
+            (click)="sage300RefreshDimensions()"
+            class="tw-h-32-px tw-w-80-px tw-mr-12-px tw-flex tw-items-center focus:tw-ring-0 tw-justify-center tw-rounded tw-bg-white tw-text-black tw-border tw-border-gray-300 focus:tw-outline-none focus:tw-ring-opacity-50 active:tw-scale-95 transition-transform">
+            <img src="assets/icons/stepper/sync.svg">
+            <span class="tw-text-14-px tw-pl-8-px">
+                Sync
+            </span>
+        </button>
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-step-header/configuration-step-header.component.html:36]: let's use existing block

File: src/app/shared/components/configuration/configuration-step-header/configuration-step-header.component.ts [hunk 6]
```diff
@@ -31,6 +35,11 @@ export class ConfigurationStepHeaderComponent implements OnInit {
     this.toastService.displayToastMessage(ToastSeverity.SUCCESS, 'Refreshing data dimensions from Sage Intacct');
   }
 
+  // This funtion name is for time being
+  sage300RefreshDimensions() {
+    this.refreshDimension.emit(true);
```
Comment (ashwin1111) [inline src/app/shared/components/configuration/configuration-step-header/configuration-step-header.component.ts:40]: can update intacct code and do it in a single place

File: src/app/core/models/sage300/db/sage300-destination-attribuite.model.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    id?: number | null;
+    attribute_type?: string;
+    display_name?: string;
+    value?: string;
+    destination_id?: string;
+    auto_created?: boolean;
+    active?: boolean | null;
+    created_at?: Date;
+    updated_at?: Date;
+    workspace?: number;
```
Comment (ashwin1111) [inline src/app/core/models/sage300/db/sage300-destination-attribuite.model.ts:15]: except detail, rest all keys would be same for every apps, move it out common. Can extend the common type and add detail only

File: src/app/core/models/sage300/db/sage300-destination-attribuite.model.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+export type EmployeeMappingDetail = {
```
Comment (ashwin1111) [inline src/app/core/models/sage300/db/sage300-destination-attribuite.model.ts:1]: if it's not used outside, need not export it

File: src/app/core/services/sage300/sage300-configuration/sage300-export-setting.service.ts [hunk 9]
```diff
@@ -38,6 +39,13 @@ export class Sage300ExportSettingService {
     return this.apiService.put(`/workspaces/${this.workspaceService.getWorkspaceId()}/export_settings/`, exportSettingsPayload);
   }
 
+  getDestinationAttributes(attributeType: string[] | string): Observable<Sage300DestinationAttributes> {
+    const params: any = {
```
Comment (ashwin1111) [inline src/app/core/services/sage300/sage300-configuration/sage300-export-setting.service.ts:43]: remove any

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html [hunk 10]
```diff
@@ -99,75 +101,52 @@
...
-                            [destinationAttributes]="sageIntacctDefaultGLAccounts"
-                            [iconPath]="'expense'"
-                            [placeholder]="'Select GL account'"
-                            [formControllerName]="'creditCard'"></app-configuration-select-field>
+                                [form]="exportSettingForm"
+                                [isFieldMandatory]="true"
+                                [mandatoryErrorListName]="'Default Credit Account Name'"
+                                [label]="'Set the Default Credit Account as?'"
+                                [subLabel]="'The integration will credit the account selected here for Corporate Card Expenses exported as Direct Cost.'"
+                                [options]="expensesExportTypeOptions"
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.html:121]: option is wrong

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 11]
```diff
@@ -43,40 +46,56 @@ export class Sage300ExportSettingsComponent implements OnInit {
 
   cccExpenseState: Sage300ExportSettingFormOption[] = this.exportSettingService.getCCCExpenseState();
 
+  validatorRule: ValidatorRule = {
+    'reimbursableExpense': ['reimbursableExportType', 'reimbursableExportGroup', 'reimbursableExportDate', 'reimbursableExpenseState'],
+    'creditCardExpense': ['cccExportType', 'cccExportGroup', 'cccExportDate', 'cccExpenseState']
+  };
+
+  defaultVendor: Sage300DestinationAttributes;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:54]: vendorOptions

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 12]
```diff
@@ -43,40 +46,56 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+
+  defaultCreditCardAccounts: Sage300DestinationAttributes;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:56]: creditCardAccountOptions

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 13]
```diff
@@ -43,40 +46,56 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+  }
 
   private setupPage(): void {
     this.isOnboarding = this.router.url.includes('onboarding');
-    this.exportSettingService.getSage300ExportSettings().subscribe((exportSettingsResponse: Sage300ExportSettingGet) => {
-      this.exportSettings = exportSettingsResponse;
+    forkJoin([
+      this.exportSettingService.getSage300ExportSettings(),
+      this.exportSettingService.getDestinationAttributes('VENDOR'),
+      this.exportSettingService.getDestinationAttributes('ACCOUNT')
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:82]: use enum for VENDOR, ACCOUNT

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 14]
```diff
@@ -43,40 +46,56 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+  save() {
+    // Will be added here soon
+  }
 
   private setupPage(): void {
     this.isOnboarding = this.router.url.includes('onboarding');
-    this.exportSettingService.getSage300ExportSettings().subscribe((exportSettingsResponse: Sage300ExportSettingGet) => {
-      this.exportSettings = exportSettingsResponse;
+    forkJoin([
+      this.exportSettingService.getSage300ExportSettings(),
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:80]: this one can fail, we cannot use this inside forkjoin directly, check if we can do .pipe(catchError(() => of(null))) to handle error

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 15]
```diff
@@ -76,20 +75,18 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
-      this.exportSettings.credit_card_expense_grouped_by = ['expense_id'];
-      this.exportSettings.reimbursable_expense_grouped_by = ['expense_id'];
-      this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup(this.exportSettings);
-      this.exportSettingForm.controls.reimbursableExpense.setValidators(this.helper.exportSelectionValidator(this.exportSettingForm));
-      this.helper.setCustomValidatorsAndWatchers(this.validatorRule, this.exportSettingForm);
-      this.defaultVendor = response[1];
-      this.defaultCreditCardAccounts = response[2];
-      this.isLoading = false;
+    this.exportSettingService.getSage300ExportSettings().subscribe((Sage300ExportSettingResponse: Sage300ExportSettingGet) => {
+      this.exportSettingService.getDestinationAttributes([FyleField.VENDOR, Sage300Field.ACCOUNT]).subscribe((response: Sage300DestinationAttributes[]) => {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:79]: use getGroupedDestinationAttributes

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 16]
```diff
@@ -76,20 +75,18 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+        this.exportSettings = Sage300ExportSettingResponse;
+        this.exportSettings.credit_card_expense_grouped_by = ['expense_id'];
+        this.exportSettings.reimbursable_expense_grouped_by = ['expense_id'];
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:82]: remove pls

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 17]
```diff
@@ -76,20 +75,18 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
+        this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup(this.exportSettings);
+        this.exportSettingForm.controls.reimbursableExpense.setValidators(this.helper.exportSelectionValidator(this.exportSettingForm));
+        this.helper.setCustomValidatorsAndWatchers(this.validatorRule, this.exportSettingForm);
+        this.vendorOptions = response;
+        this.creditCardAccountOptions = response;
+        this.isLoading = false;
+      });
     }, (error) => {
       this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup();
       this.isLoading = false;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:92]: this.helper.setCustomValidatorsAndWatchers(this.validatorRule, this.exportSettingForm);

File: src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts [hunk 18]
```diff
@@ -76,20 +75,18 @@ export class Sage300ExportSettingsComponent implements OnInit {
...
-      this.exportSettings = response[0];
-      this.exportSettings.credit_card_expense_grouped_by = ['expense_id'];
-      this.exportSettings.reimbursable_expense_grouped_by = ['expense_id'];
-      this.exportSettingForm = ExportSettingModel.mapAPIResponseToFormGroup(this.exportSettings);
-      this.exportSettingForm.controls.reimbursableExpense.setValidators(this.helper.exportSelectionValidator(this.exportSettingForm));
-      this.helper.setCustomValidatorsAndWatchers(this.validatorRule, this.exportSettingForm);
-      this.defaultVendor = response[1];
-      this.defaultCreditCardAccounts = response[2];
-      this.isLoading = false;
+    this.exportSettingService.getSage300ExportSettings().subscribe((Sage300ExportSettingResponse: Sage300ExportSettingGet) => {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-shared/sage300-export-settings/sage300-export-settings.component.ts:78]: https://github.com/fylein/fyle-integrations-settings-app/pull/224#discussion_r1372967388

**Comment by DhaaraniCIT** [comment]
This PR comments are fixed in branch 'export-setting-save' PR: [https://github.com/fylein/fyle-integrations-settings-app/pull/233](url)

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**